### PR TITLE
Deduplicate incompatible `target_compatible_with` labels

### PR DESCRIPTION
**Pull request by philsc**
The latest refactor unintentionally made it so you can list the same
incompatible constraint multiple times in the `target_compatible_with`
attribute.

It was unintentional, but actually greatly simplifies a discussion
point on bazelbuild/bazel-skylib#381. That skylib patch aims to make
it easier to write non-trivial `target_compatible_with` expressions.
For example, to express that something is compatible with everything
except for Windows, one can use the following:

    foo_binary(
        name = "bin",
        target_compatible_with = select({
            "@platforms//os:windows": ["@platforms//:incomaptible"],
            "//conditions:default: [],
        }),
    )

The skylib patch aims to reduce that to:

    foo_binary(
        name = "bin",
        target_compatible_with = compatibility.none_of("@platforms//os:windows"),
    )

This works fine on its own, but a problem arises when these
expressions are composed. For example, a macro author might write the
following:

    def foo_wrapped_binary(name, target_compatible_with = [], **kwargs):
        foo_binary(
            name = name,
            target_compatible_with = target_compatible_with + select({
                "@platforms//os:none": ["@platforms//:incompatible"],
                "//conditions:default": [],
            }),
        )

A macro author should be able to express their own incompatibility
while also honouring user-defined incompatibility.

It turns out that my latest refactor (#14096) unintentionally made
this work. This happened because we now check for incompatibility
before we perform a lot of sanity checks. That's also one of the
reasons that visibility restrictions are not honoured for incomaptible
targets at the moment (#16044).

Without the deduplicating behaviour, macro authors are stuck with
either not allowing composition, or having to create unique
incompatible constraints for every piece in a composed
`target_compatible_with` expression.

This patch adds a test to validate the deduplicating behaviour to
cement it as a feature.

Unfortunately, this means that `target_compatible_with` behaves
differently from other label list attributes. For other label list
attributes, bazel complains when labels are duplicated. For example,
the following BUILD file:

    py_library(
        name = "lib",
    )

    py_binary(
        name = "bin",
        srcs = ["bin.py"],
        deps = [
        |   ":lib",
        |   ":lib",
        ],
    )

will result in the following error:

    $ bazel build :bin
    INFO: Invocation ID: 3d8345a4-2e76-493e-8343-c96a36185b52
    ERROR: /home/jenkins/repos/test_repo/BUILD:41:10: Label '//:lib' is duplicated in the 'deps' attribute of rule 'bin'
    ERROR: error loading package '': Package '' contains errors
    INFO: Elapsed time: 2.514s
    INFO: 0 processes.
    FAILED: Build did NOT complete successfully (1 packages loaded)

### CODE REVIEW

**Comment by philsc** [comment]
@gregestren , @katre , at the risk of disappointing you with my poor decision making, what do you think of this patch?

File: src/main/java/com/google/devtools/build/lib/analysis/IncompatiblePlatformProvider.java [hunk 1]
```diff
@@ -95,6 +105,8 @@ public boolean isImmutable() {
    *
    * <p>This may be null. If it is null, then {@code getTargetsResponsibleForIncompatibility()} is
    * guaranteed to be non-null. It will have at least one element in it if it is not null.
+   *
+   * <p>The list is sorted based on the toString() result of each constraint.
```
Comment (philsc) [inline src/main/java/com/google/devtools/build/lib/analysis/IncompatiblePlatformProvider.java:109]: Probably just reword this to be "stringified label" rather than "toString()".

File: src/main/java/com/google/devtools/build/lib/analysis/IncompatiblePlatformProvider.java [hunk 2]
```diff
@@ -69,7 +70,16 @@ public static IncompatiblePlatformProvider incompatibleDueToConstraints(
       @Nullable Label targetPlatform, ImmutableList<ConstraintValueInfo> constraints) {
     Preconditions.checkNotNull(constraints);
     Preconditions.checkArgument(!constraints.isEmpty());
-    return new AutoValue_IncompatiblePlatformProvider(targetPlatform, null, constraints);
+
+    // Deduplicate and sort the list of constraints. Doing it here means that everyone inspecting
+    // this provider doesn't have to deal with it.
+    ImmutableSet<ConstraintValueInfo> filteredConstraints = ImmutableSet.copyOf(constraints);
```
Comment (katre) [inline src/main/java/com/google/devtools/build/lib/analysis/IncompatiblePlatformProvider.java:76]: It'd be easier to do this in one step (and to re-use the existing name, so that the old data can't be referenced):

```
constraints = constreams.stream()
    .sorted(Comparator.comparing(ConstraintValueInfo::label))
    .distinct()
    .collect(toImmutableList());

File: src/main/java/com/google/devtools/build/lib/analysis/IncompatiblePlatformProvider.java (hunk 1 above)
Comment (katre) [reply src/main/java/com/google/devtools/build/lib/analysis/IncompatiblePlatformProvider.java:109]: +1

**Review by katre — CHANGES_REQUESTED** [review]
I think this makes a lot of sense, and there are probably more places in Bazel where we should deduplicate instead of erroring.

**Comment by gregestren** [comment]
https://github.com/bazelbuild/bazel/commit/2570fe1119b2462ba10790d3bedf53699baecdf5 is a recent change recognizing some subtlety with `select()`. It's not the same purpose as yours, but a reminder that dedupe logic can't be exactly the same everwyhere.

This all reads reasonable to me but my main concern is API consistency. Surely this kind of theme has come up before w.r.t. macros? Let me run this by @comius and/or @brandjon , both of who I'll chat with tomorrow.

File: src/main/java/com/google/devtools/build/lib/analysis/IncompatiblePlatformProvider.java (hunk 1 above)
Comment (philsc) [reply src/main/java/com/google/devtools/build/lib/analysis/IncompatiblePlatformProvider.java]: Done

File: src/main/java/com/google/devtools/build/lib/analysis/IncompatiblePlatformProvider.java (hunk 2 above)
Comment (philsc) [reply src/main/java/com/google/devtools/build/lib/analysis/IncompatiblePlatformProvider.java]: Done

**Review by katre — APPROVED** [review]

**Comment by katre** [comment]
Looks good to me, waitinf ro @gregestren, @comius, and @brandjon to weigh in.

**Review by comius — APPROVED** [review]

**Review by gregestren — APPROVED** [review]
Sounds like no concerns.
